File: C#/Perfect Numbers/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    static List<int> Factors(int n)
    {
        List<int> factors = [];

        for (int i = 1; i < n; i++)
        {
            if (n % i == 0)
            {
                factors.Add(i);
            }
        }

        return factors;
    }

    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException("Classification is only possible for natural numbers.");
        }

        var factorsSum = Factors(number).Sum();

        if(factorsSum == number)
        {
            return Classification.Perfect;
        }
        else if(factorsSum > number)
        {
            return Classification.Abundant;
        }

        return Classification.Deficient;
    }
}

File: TypeScript/perfect-numbers.ts
const factors = (n:number) :number[] => {
    let factors = [];

    for (let i = 1; i < n; i++) {
        if(n % i == 0) {
            factors.push(i);
        }
    }

    return factors;
};

function add(accumulator:number, a:number) :number {
    return accumulator + a;
};

export const classify = (n: number): string => {
    if(n <= 0) {
        throw new Error('Classification is only possible for natural numbers.')
    }

    let f = factors(n);
    let s = f.reduce(add, 0);

    if(s === n) {
        return 'perfect';
    }

    else if(s > n) {
        return 'abundant';
    }

    //else if(s < n)
    return 'deficient';
};

const numbers = [
    28,
    12,
    33550335,
    32,
]

numbers.forEach(n => {
    console.log(`${n} -> ${classify(n)}`);
});